src/wasm/wasm-binary.cpp

```diff
@@ -1536,20 +1559,23 @@ void WasmBinaryBuilder::readDataSegments() {
   if (debug) std::cerr << "== readDataSegments" << std::endl;
   auto num = getU32LEB();
   for (size_t i = 0; i < num; i++) {
-    auto memoryIndex = getU32LEB();
-    WASM_UNUSED(memoryIndex);
-    if (memoryIndex != 0) {
-      throwError("bad memory index, must be 0");
-    }
     Memory::Segment curr;
-    auto offset = readExpression();
+    curr.flags = getU32LEB();
+    if (curr.flags > 2) {
+      throwError("bad segment flags, must be 0, 1, or 2, not " + std::to_string(curr.flags));
+    }
+    if (curr.hasMemIndex()) {
+      curr.index = getU32LEB();
+    }
+    if (!curr.isPassive()) {
+      curr.offset = readExpression();
+    }
     auto size = getU32LEB();
-    std::vector<char> buffer;
-    buffer.resize(size);
+    curr.data.resize(size);
     for (size_t j = 0; j < size; j++) {
-      buffer[j] = char(getInt8());
+      curr.data[j] = char(getInt8());
     }
-    wasm.memory.segments.emplace_back(offset, (const char*)&buffer[0], size);
+    wasm.memory.segments.push_back(curr);
```

Comment:
this `push_back` causes an extra copy, I think, compared to before?

Reply:
Technically yes, but I think it is more readable when written this way due to all the conditionally meaningful fields, and who knows what the optimizer will do with it anyway.